File: quant1x/data/meta/frequency.go
// Copyright (c) Quant1X <wangfengxy@sina.cn>.
// Licensed under the MIT License.

package meta

import (
	"errors"
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

// TimeUnit 标准化的时间单位枚举，覆盖 pandas 常见别名。
// 所有单位均为固定长度（不包括月、年等日历单位）。
type TimeUnit string

const (
	NANOSECOND  TimeUnit = "ns"
	MICROSECOND TimeUnit = "us"
	MILLISECOND TimeUnit = "ms"
	SECOND      TimeUnit = "s"
	MINUTE      TimeUnit = "min"
	HOUR        TimeUnit = "h"
	DAY         TimeUnit = "d"
	WEEK        TimeUnit = "w"
	MONTH       TimeUnit = "m"
	YEAR        TimeUnit = "y"
)

// SecondsPerUnit 每单位对应的秒数（float64 支持纳秒）
var SecondsPerUnit = map[TimeUnit]float64{
	NANOSECOND:  1e-9,
	MICROSECOND: 1e-6,
	MILLISECOND: 1e-3,
	SECOND:      1.0,
	MINUTE:      60.0,
	HOUR:        3600.0,
	DAY:         86400.0,
	WEEK:        604800.0,
	MONTH:       2592000.0,
	YEAR:        31536000.0,
}

// Frequency 表示一个标准化的频率值: num x unit。
// 例如: 5 分钟 → Frequency{Num: 5, Unit: MINUTE}
type Frequency struct {
	Num  int
	Unit TimeUnit
}

// ToTotalSeconds 返回总秒数（可用于比较、排序、计算）
func (f Frequency) ToTotalSeconds() float64 {
	return float64(f.Num) * SecondsPerUnit[f.Unit]
}

// String 返回频率字符串表示
func (f Frequency) String() string {
	return fmt.Sprintf("%d%s", f.Num, f.Unit)
}

// CacheKey 返回缓存键
func (f Frequency) CacheKey() string {
	if f.Unit == DAY {
		return "day"
	}
	return fmt.Sprintf("%d%s", f.Num, f.Unit)
}

// 预定义的频率常量
var (
	FREQ_DAILY  = Frequency{Num: 1, Unit: DAY}   // 日线
	FREQ_WEEKLY = Frequency{Num: 1, Unit: WEEK}  // 周线
	FREQ_MONTHLY = Frequency{Num: 1, Unit: MONTH} // 月线
	FREQ_YEARLY  = Frequency{Num: 1, Unit: YEAR}  // 年线
)

// pandas 单位别名映射表
var pandasUnitAliases = map[string]TimeUnit{
	// nanosecond
	"N":  NANOSECOND,
	"ns": NANOSECOND,
	// microsecond
	"U":  MICROSECOND,
	"us": MICROSECOND,
	"µs": MICROSECOND,
	// millisecond
	"L":  MILLISECOND,
	"ms": MILLISECOND,
	// second
	"S": SECOND,
	"s": SECOND,
	// minute
	"T": MINUTE,
	"min": MINUTE,
	// hour
	"H": HOUR,
	"h": HOUR,
	// day
	"D": DAY,
	"d": DAY,
	// week
	"W": WEEK,
	"w": WEEK,
	// month
	"M": MONTH,
	"m": MONTH,
	// year
	"Y": YEAR,
	"y": YEAR,
}

// ParseFrequencyString 解析 pandas 风格的频率字符串（如 '5T', '1H', '30s'）为标准化 Frequency。
//
// 参数:
//   freq: 频率字符串，如 "5T", "1h", "90s"
//
// 返回:
//   Frequency{Num: 5, Unit: MINUTE}
//
// 错误:
//   无效格式或不支持的单位
func ParseFrequencyString(freq string) (Frequency, error) {
	s := strings.TrimSpace(freq)
	if s == "" {
		return Frequency{}, errors.New("frequency string is empty")
	}

	// 正则提取数字前缀和单位后缀
	re := regexp.MustCompile(`^(\d*)(.*)$`)
	match := re.FindStringSubmatch(s)
	if match == nil {
		return Frequency{}, fmt.Errorf("invalid frequency format: %s", s)
	}

	numStr := match[1]
	unitStr := match[2]

	num := 1
	if numStr != "" {
		var err error
		num, err = strconv.Atoi(numStr)
		if err != nil {
			return Frequency{}, fmt.Errorf("invalid number format: %s", numStr)
		}
	}

	if unitStr == "" {
		return Frequency{}, errors.New("missing unit in frequency string")
	}

	unit, ok := pandasUnitAliases[unitStr]
	if !ok {
		return Frequency{}, fmt.Errorf("unsupported or unknown frequency unit: %q", unitStr)
	}

	return Frequency{Num: num, Unit: unit}, nil
}

// ToTotalSeconds 便捷函数：将频率转为总秒数
func ToTotalSeconds(freq interface{}) (float64, error) {
	var f Frequency
	var err error

	switch v := freq.(type) {
	case string:
		f, err = ParseFrequencyString(v)
		if err != nil {
			return 0, err
		}
	case Frequency:
		f = v
	default:
		return 0, errors.New("invalid frequency type: must be string or Frequency")
	}

	return f.ToTotalSeconds(), nil
}

// IsFixedDuration 判断是否为固定时长（所有当前支持的单位都是固定的）。
// 未来若加入 'M'（月）、'Y'（年），此处需调整。
func IsFixedDuration(freq interface{}) bool {
	// 当前所有单位均为固定长度
	return true
}

// ToFrequency 便捷函数：将字符串转换为 Frequency
func ToFrequency(freq string) Frequency {
	f, err := ParseFrequencyString(freq)
	if err != nil {
		panic(err)
	}
	return f
}

File: quant1x/data/meta/frequency.py
# -*- coding: utf-8 -*-
# Copyright (c) Quant1X <wangfengxy@sina.cn>.
# Licensed under the MIT License.

import re
from enum import Enum
from typing import NamedTuple, Union

__all__ = [
    "TimeUnit",
    "Frequency",
    "parse_frequency_string",
    "is_fixed_duration",
    "to_total_seconds",
]


class TimeUnit(Enum):
    """
    标准化的时间单位枚举, 覆盖 pandas 常见别名. 
    所有单位均为固定长度(不包括月, 年等日历单位). 
    """
    NANOSECOND = "ns"
    MICROSECOND = "us"
    MILLISECOND = "ms"
    SECOND = "s"
    MINUTE = "min"
    HOUR = "h"
    DAY = "d"
    WEEK = "w"
    MONTH = "m"
    YEAR = "y"

    @property
    def seconds_per_unit(self) -> float:
        """每单位对应的秒数(float 支持纳秒)"""
        mapping = {
            TimeUnit.NANOSECOND: 1e-9,
            TimeUnit.MICROSECOND: 1e-6,
            TimeUnit.MILLISECOND: 1e-3,
            TimeUnit.SECOND: 1.0,
            TimeUnit.MINUTE: 60.0,
            TimeUnit.HOUR: 3600.0,
            TimeUnit.DAY: 86400.0,
            TimeUnit.WEEK: 604800.0,
            TimeUnit.MONTH: 2592000.0,
            TimeUnit.YEAR: 31536000.0,
        }
        return mapping[self]


class Frequency(NamedTuple):
    """
    表示一个标准化的频率值: num x unit. 
    例如: 5 分钟 → Frequency(num=5, unit=TimeUnit.MINUTE)
    """
    num: int
    unit: TimeUnit

    def to_total_seconds(self) -> float:
        """返回总秒数(可用于比较, 排序, 计算)"""
        return self.num * self.unit.seconds_per_unit

    def __str__(self) -> str:
        return f"{self.num}{self.unit.value}"
    
    def cache_key(self) -> str:
        if self.unit == TimeUnit.DAY:
            return 'day'
        return f"{self.num}{self.unit.value}"

# 日线频率的常量
FREQ_DAILY = Frequency(num=1, unit=TimeUnit.DAY)
"""日线"""
FREQ_WEEKLY = Frequency(num=1, unit=TimeUnit.WEEK)
"""周线"""
FREQ_MONTHLY = Frequency(num=1, unit=TimeUnit.MONTH)
"""月线"""
FREQ_YEARLY = Frequency(num=1, unit=TimeUnit.YEAR)
"""年线"""

# pandas 单位别名映射表(只读)
_PANDAS_UNIT_ALIASES = {
    # nanosecond
    "N": TimeUnit.NANOSECOND, "ns": TimeUnit.NANOSECOND,
    # microsecond
    "U": TimeUnit.MICROSECOND, "us": TimeUnit.MICROSECOND, "µs": TimeUnit.MICROSECOND,
    # millisecond
    "L": TimeUnit.MILLISECOND, "ms": TimeUnit.MILLISECOND,
    # second
    "S": TimeUnit.SECOND, "s": TimeUnit.SECOND,
    # minute
    "T": TimeUnit.MINUTE, "min": TimeUnit.MINUTE,
    # hour
    "H": TimeUnit.HOUR, "h": TimeUnit.HOUR,
    # day
    "D": TimeUnit.DAY, "d": TimeUnit.DAY,
    # week
    "W": TimeUnit.WEEK, "w": TimeUnit.WEEK,
    # month
    "M": TimeUnit.MONTH, "m": TimeUnit.MONTH,
    # year
    "Y": TimeUnit.YEAR, "y": TimeUnit.YEAR,
}

def parse_frequency_string(freq: str) -> Frequency:
    """
    解析 pandas 风格的频率字符串(如 '5T', '1H', '30s')为标准化 FrequencyValue. 
    
    Args:
        freq: 频率字符串, 如 "5T", "1h", "90s"
    
    Returns:
        Frequency(count=5, unit=TimeUnit.MINUTE)
    
    Raises:
        ValueError: 无效格式或不支持的单位
    """
    s = freq.strip()
    if not s:
        raise ValueError("frequency string is empty")

    # 正则提取数字前缀和单位后缀
    match = re.match(r'^(\d*)(.*)$', s)
    if not match:
        raise ValueError(f"invalid frequency format: {s}")

    num_str, unit_str = match.groups()
    num = int(num_str) if num_str else 1

    if not unit_str:
        raise ValueError("missing unit in frequency string")

    unit = _PANDAS_UNIT_ALIASES.get(unit_str)
    if unit is None:
        raise ValueError(f"unsupported or unknown frequency unit: {unit_str!r}")

    return Frequency(num=num, unit=unit)


def to_total_seconds(freq: Union[str, Frequency]) -> float:
    """便捷函数: 将频率转为总秒数"""
    if isinstance(freq, str):
        freq = parse_frequency_string(freq)
    return freq.to_total_seconds()


def is_fixed_duration(freq: Union[str, Frequency]) -> bool:
    """
    判断是否为固定时长(所有当前支持的单位都是固定的). 
    未来若加入 'M'(月), 'Y'(年), 此处需调整. 
    """
    return True  # 当前所有单位均为固定长度

if __name__ == "__main__":
    f1 = parse_frequency_string("5T")
    print(f1)  # 5min
    print(f1.count, f1.unit)  # 5 TimeUnit.MINUTE
    print(f1.to_total_seconds())  # 300.0

    f2 = parse_frequency_string("1h")
    print(f2.to_total_seconds())  # 3600.0

    # 可用于比较
    assert f1.to_total_seconds() < f2.to_total_seconds()